File: chassis-workbench/src/engine/geometry.ts
/**
 * geometry.ts — Chassis Geometry Engine
 *
 * Implements Equations 5.1–5.5 from the Motorcycle Chassis Dynamics
 * Workbench Technical Specification v3.0.
 *
 * PRIMARY REFERENCE:
 *   Foale, T. (2006). Motorcycle Handling and Chassis Design, Chapter 2.
 *   Tony Foale Designs. ISBN 978-84-933286-3-4
 *
 * SECONDARY REFERENCES:
 *   Cossalter, V. (2006). Motorcycle Dynamics, 2nd Ed., Chapter 1.
 *   SAE J1168 — Motorcycle Terminology
 *
 * NOTE ON EQ 5.1 (TRAIL FORMULA):
 *   The spec document lists: Trail = (R_f × cos α) / sin α − f
 *   This form is valid only when α is measured from the HORIZONTAL
 *   (the convention Foale uses internally in his derivation diagrams).
 *   However, the spec's own input parameter definition (Section 5.1)
 *   defines headAngle as "angle of steering axis from VERTICAL".
 *
 *   Using the spec formula with α-from-vertical and typical sport bike
 *   values (R_f=310mm, α=24°, f=25mm) produces Trail = 671mm —
 *   physically impossible for any road motorcycle.
 *
 *   The CORRECT implementation for α-from-vertical is derived from
 *   first principles (and confirmed by Foale's worked examples):
 *
 *       Trail = (R_f × sin α − f) / cos α        ← IMPLEMENTED HERE
 *
 *   This is Option B as agreed with the user. For Yamaha R1 parameters
 *   (R_f≈300mm, α=24°, f=25mm) it yields Trail ≈ 106mm, consistent
 *   with Yamaha's published specification of 97mm (the 9mm difference
 *   is the loaded vs free-body tyre radius under rider weight).
 *
 * UNITS: mm for all lengths, radians internally, degrees in public API.
 */

import { GeometryParams, GeometryResults } from './types';

const DEG2RAD = Math.PI / 180;
const RAD2DEG = 180 / Math.PI;

// ─────────────────────────────────────────────────────────
// PURE FUNCTIONS
// ─────────────────────────────────────────────────────────

/**
 * Computes geometric trail.
 *
 * Trail is the horizontal distance from the front tyre contact patch to
 * the point where the steering axis, extended downward, intersects the
 * ground. Positive trail means the contact patch trails BEHIND the axis
 * intersection — the normal, self-stabilising condition.
 *
 * Trail = (R_f × sin α − f) / cos α           ... Eq 5.1 (Foale Ch.2)
 *
 * @param R_f        Front tyre radius (mm) = frontWheelDia / 2
 * @param alpha_deg  Head angle from VERTICAL (degrees), typical 22–32°
 * @param forkOffset Perpendicular distance from steering axis to front
 *                   axle centre (mm), typical 22–45mm
 * @returns Trail in mm. Positive = stable. Negative = dangerously unstable.
 * @throws  RangeError if cos α ≈ 0 (fork horizontal — undefined geometry)
 */
export function computeTrail(
  R_f: number,
  alpha_deg: number,
  forkOffset: number,
): number {
  const alpha = alpha_deg * DEG2RAD;
  const cosA = Math.cos(alpha);

  if (Math.abs(cosA) < 1e-9) {
    throw new RangeError(
      `computeTrail: cos(${alpha_deg}°) ≈ 0. ` +
      `A horizontal steering axis produces undefined trail geometry.`,
    );
  }

  // Eq 5.1: Trail = (R_f · sin α − f) / cos α
  // Derivation: steering axis hits ground at x = (R_f·sin α − f)/cos α
  // ahead of the contact patch (positive x = forward = stable).
  return (R_f * Math.sin(alpha) - forkOffset) / cosA;
}

/**
 * Computes mechanical trail.
 *
 * Mechanical trail is the perpendicular distance from the tyre contact
 * patch to the steering axis, measured ALONG the steering axis direction
 * (not horizontally). It is the actual lever arm producing self-aligning
 * torque (SAT). Always larger than geometric trail for α > 0.
 *
 * MechanicalTrail = Trail / cos α              ... Eq 5.2 (Foale Ch.2)
 *
 * @param trail      Geometric trail (mm) from computeTrail()
 * @param alpha_deg  Head angle from vertical (degrees)
 * @returns Mechanical trail in mm
 */
export function computeMechanicalTrail(
  trail: number,
  alpha_deg: number,
): number {
  const alpha = alpha_deg * DEG2RAD;
  const cosA = Math.cos(alpha);

  if (Math.abs(cosA) < 1e-9) {
    throw new RangeError(
      `computeMechanicalTrail: cos(${alpha_deg}°) ≈ 0.`,
    );
  }

  // Eq 5.2
  return trail / cosA;
}

/**
 * Computes the fork offset projected horizontally to the ground plane.
 *
 * This is the effective horizontal displacement the fork offset creates
 * at ground level. Used to understand how offset changes the ground-
 * level steering axis position relative to a zero-offset fork.
 *
 * SteeringOffset_ground = f × cos α           ... Eq 5.3
 *
 * @param forkOffset  Fork offset f (mm)
 * @param alpha_deg   Head angle from vertical (degrees)
 * @returns Ground-level steering offset in mm
 */
export function computeSteeringOffsetGround(
  forkOffset: number,
  alpha_deg: number,
): number {
  const alpha = alpha_deg * DEG2RAD;
  // Eq 5.3
  return forkOffset * Math.cos(alpha);
}

/**
 * Computes the static swingarm angle relative to horizontal.
 *
 * θ_sa = arcsin((H_ra − H_sp) / L_sa)         ... Eq 5.4 (corrected)
 *
 * L_sa is the total swingarm length (hypotenuse of the pivot-to-axle triangle).
 * Using atan(Δy / L_sa) is WRONG because it treats the hypotenuse as the
 * adjacent side.  The correct formula uses arcsin (or equivalently
 * atan2(Δy, √(L_sa²−Δy²))).
 *
 * Sign convention:
 *   Negative → rear axle is LOWER than pivot (typical — swingarm slopes
 *              downward toward wheel). Supermoto: −1° to −5°.
 *   Positive → rear axle is HIGHER than pivot (rare, some off-road bikes).
 *
 * Realistic supermoto ranges:
 *   Pivot height H_sp: 320–360 mm
 *   Rear axle height H_ra: 285–315 mm
 *   Resulting θ_sa: −1° to −5° (NOT −15°)
 *
 * @param H_ra  Rear axle centre height above ground (mm)
 * @param H_sp  Swingarm pivot height above ground (mm)
 * @param L_sa  Swingarm length, pivot centre-to-axle centre (mm)
 * @returns Swingarm angle in radians
 * @throws  RangeError if L_sa ≈ 0 or geometry is impossible (|Δy| > L_sa)
 */
export function computeSwingarmAngle(
  H_ra: number,
  H_sp: number,
  L_sa: number,
): number {
  if (Math.abs(L_sa) < 1e-9) {
    throw new RangeError(
      `computeSwingarmAngle: L_sa ≈ 0. Swingarm length cannot be zero.`,
    );
  }
  const deltaY = H_ra - H_sp;
  if (Math.abs(deltaY) > L_sa) {
    throw new RangeError(
      `computeSwingarmAngle: |H_ra − H_sp| (${Math.abs(deltaY).toFixed(1)} mm) ` +
      `exceeds swingarm length L_sa (${L_sa.toFixed(1)} mm). ` +
      `Check pivot height (supermoto: 320–360 mm) and rear axle height (285–315 mm).`,
    );
  }
  // Eq 5.4 (corrected): L_sa is the hypotenuse, so use arcsin, NOT arctan.
  // arctan(Δy / L_sa) underestimates |θ| at all angles > 0.
  return Math.asin(deltaY / L_sa);
}

/**
 * Computes self-aligning torque about the steering axis.
 *
 * M_sc = F_N_front × Trail × sin(δ)           ... Eq 5.5
 *
 * This is the moment (about the steering axis) produced by the front
 * tyre's normal reaction acting through the trail moment arm. It is
 * what makes the handlebar return to centre when released.
 *
 * Trail is converted internally from mm to metres so the result is N·m.
 *
 * @param F_N_front      Normal load on front wheel (N) from static or
 *                       dynamic analysis
 * @param trail_mm       Geometric trail (mm) from computeTrail()
 * @param steerAngle_deg Steering (steer) angle δ (degrees)
 * @returns Self-aligning torque (N·m). Positive = restoring force.
 */
export function computeSelfAligningTorque(
  F_N_front: number,
  trail_mm: number,
  steerAngle_deg: number,
): number {
  const delta = steerAngle_deg * DEG2RAD;
  const trail_m = trail_mm / 1000; // mm → m for N·m output
  // Eq 5.5: M_sc = F_N_front × Trail[m] × sin(δ)
  return F_N_front * trail_m * Math.sin(delta);
}

// ─────────────────────────────────────────────────────────
// AGGREGATE FUNCTION
// ─────────────────────────────────────────────────────────

/**
 * Computes all chassis geometry derived values from the 18-parameter
 * GeometryParams struct.
 *
 * This is the only function that should be called by computeAll.ts.
 * All individual functions above remain independently testable.
 *
 * @param p  GeometryParams — all angles in degrees, all lengths in mm
 * @returns  GeometryResults
 */
export function computeGeometry(p: GeometryParams): GeometryResults {
  const R_f = p.frontWheelDia / 2;
  const R_r = p.rearWheelDia / 2;

  const trail = computeTrail(R_f, p.headAngle, p.forkOffset);
  const mechanicalTrail = computeMechanicalTrail(trail, p.headAngle);
  const steeringOffsetGround = computeSteeringOffsetGround(
    p.forkOffset,
    p.headAngle,
  );

  // Swingarm angle: atan2(H_ra − H_sp, WB − X_sp) uses the actual pivot-to-axle
  // vector, so the angle is exactly consistent with the rear axle position stored
  // in the state. computeAntiSquatUnified uses the same formula, ensuring the
  // displayed SA angle matches the angle used for IC / AS% construction.
  //
  // Note: computeSwingarmAngle(H_ra, H_sp, L_sa) uses asin(ΔY/L_sa) which is
  // correct for kinematic sweep (swingarm arc), but diverges from atan2 when
  // the stored wheelbase is not derived from L_sa (independent inputs). The
  // static display should match the pivot-to-axle direction, not the arc formula.
  const deltaX = p.wheelbase - p.swingarmPivotX;
  const deltaY = p.rearAxleHeight - p.swingarmPivotHeight;
  const swingarmAngleRad = Math.atan2(deltaY, deltaX);

  return {
    trail,
    mechanicalTrail,
    steeringOffsetGround,
    swingarmAngleRad,
    swingarmAngleDeg: swingarmAngleRad * RAD2DEG,
    frontWheelRadius: R_f,
    rearWheelRadius: R_r,
  };
}

File: chassis_sim/geometry.py
"""
chassis_sim/geometry.py — Static Chassis Geometry Engine

Computes all static geometry outputs (trail, swingarm angle, anti-squat %,
weight distribution, CoG) from a BikeGeometry input.

All formulas verified against:
  Foale (2006) Motorcycle Handling and Chassis Design
  Cossalter (2006) Motorcycle Dynamics, 2nd Ed.

Units: mm for all lengths, degrees in public API, radians internally.
"""

from __future__ import annotations
import math
from dataclasses import dataclass, field
from typing import List

# ── Input dataclasses ─────────────────────────────────────────────────────────

@dataclass
class BikeGeometry:
    """Primary chassis geometry inputs."""
    head_angle_deg: float       # rake — steering axis from vertical
    fork_offset_mm: float       # perpendicular: steering axis to front axle
    front_wheel_dia_mm: float   # outer diameter incl. tyre
    rear_wheel_dia_mm: float
    wheelbase_mm: float
    swingarm_length_mm: float   # pivot centre to rear axle centre
    swingarm_pivot_height_mm: float  # H_sp — pivot height from ground
    swingarm_pivot_x_mm: float       # X_sp — pivot distance from front axle
    rear_axle_height_mm: float       # H_ra — rear axle height from ground

    @property
    def R_f(self) -> float:
        return self.front_wheel_dia_mm / 2

    @property
    def R_r(self) -> float:
        return self.rear_wheel_dia_mm / 2

    @property
    def alpha_rad(self) -> float:
        return math.radians(self.head_angle_deg)


@dataclass
class MassComponent:
    mass_kg: float
    x_mm: float   # from front axle (positive rearward)
    y_mm: float   # from ground (positive upward)
    label: str = ""


@dataclass
class ChainGeometry:
    front_sprocket_teeth: int
    rear_sprocket_teeth: int
    sprocket_center_x_mm: float  # countershaft offset from SA pivot (neg = forward)
    sprocket_center_y_mm: float  # countershaft offset from SA pivot (pos = above)
    chain_force_angle_deg: float # top chain run angle from horizontal


# ── Static geometry formulas ──────────────────────────────────────────────────

def compute_trail(geom: BikeGeometry) -> float:
    """
    Geometric trail.

    T = (R_f · sin α − f) / cos α        ... Foale Eq 2.1

    α = head angle from vertical (positive = fork leans forward).
    f = fork offset (perpendicular distance: steering axis to front axle).
    Positive trail → self-stabilising.

    Validated: sport bike R_f=300, α=24°, f=33 → T ≈ 96 mm (Yamaha R1 ≈ 97 mm).
    """
    a = geom.alpha_rad
    cos_a = math.cos(a)
    if abs(cos_a) < 1e-9:
        raise ValueError("Head angle 90° — undefined trail geometry")
    return (geom.R_f * math.sin(a) - geom.fork_offset_mm) / cos_a


def compute_mechanical_trail(trail_mm: float, head_angle_deg: float) -> float:
    """
    Mechanical trail = perpendicular lever arm for self-aligning torque.

    MT = T / cos α                        ... Foale Eq 2.2
    """
    cos_a = math.cos(math.radians(head_angle_deg))
    if abs(cos_a) < 1e-9:
        raise ValueError("cos(head_angle) ≈ 0")
    return trail_mm / cos_a


def compute_swingarm_angle_rad(geom: BikeGeometry) -> float:
    """
    Static swingarm angle from horizontal.

    θ_sa = arctan((H_ra − H_sp) / L_sa)  ... derived from pivot-to-axle vector

    Negative for typical bike (axle below pivot). Range: −3° to −8°.
    """
    if abs(geom.swingarm_length_mm) < 1e-9:
        raise ValueError("Swingarm length cannot be zero")
    return math.atan2(
        geom.rear_axle_height_mm - geom.swingarm_pivot_height_mm,
        geom.swingarm_length_mm
    )


def compute_cog(components: List[MassComponent]) -> tuple[float, float, float]:
    """
    Centre of gravity from component masses.

    X_cg = Σ(m_i · x_i) / Σm_i          ... Eq 6.1
    Y_cg = Σ(m_i · y_i) / Σm_i          ... Eq 6.2

    Returns (X_cg_mm, Y_cg_mm, total_mass_kg).
    """
    if not components:
        raise ValueError("No mass components provided")
    total = sum(c.mass_kg for c in components)
    if total < 1e-9:
        raise ValueError("Total mass is zero")
    X_cg = sum(c.mass_kg * c.x_mm for c in components) / total
    Y_cg = sum(c.mass_kg * c.y_mm for c in components) / total
    return X_cg, Y_cg, total


def compute_axle_loads(total_mass_kg: float, X_cg_mm: float, wheelbase_mm: float):
    """
    Static axle reactions from moment balance.

    R_front = W · (WB − X_cg) / WB       ... Eq 6.5
    R_rear  = W · X_cg / WB              ... Eq 6.6
    """
    G = 9.81
    if abs(wheelbase_mm) < 1e-9:
        raise ValueError("Wheelbase cannot be zero")
    W = total_mass_kg * G
    R_front = W * (wheelbase_mm - X_cg_mm) / wheelbase_mm
    R_rear  = W * X_cg_mm / wheelbase_mm
    front_pct = (wheelbase_mm - X_cg_mm) / wheelbase_mm * 100
    return R_front, R_rear, front_pct


def compute_instant_centre(
    geom: BikeGeometry,
    chain: ChainGeometry,
    swingarm_angle_rad: float,
) -> tuple[float, float]:
    """
    Instant Centre (IC) of the anti-squat geometry.

    The IC is the intersection of two lines (Foale Ch. 11):
      Line 1: Swingarm axis extended forward
      Line 2: Top chain run line (through countershaft, along chain force angle)

    Solves:
      m1 = tan(θ_sa)
      m2 = tan(θ_chain)
      IC_x = (y2_0 − y1_0 + m1·x1 − m2·x2) / (m1 − m2)
      IC_y = H_sp + m1·(IC_x − X_sp)

    Returns (IC_x_mm, IC_y_mm).
    Raises ValueError if lines are parallel (m1 ≈ m2).
    """
    m1 = math.tan(swingarm_angle_rad)
    m2 = math.tan(math.radians(chain.chain_force_angle_deg))
    if abs(m1 - m2) < 1e-9:
        raise ValueError(
            f"Swingarm and chain lines parallel (m={m1:.4f}) — no IC. "
            "Adjust chain force angle or swingarm angle."
        )
    X_sp, H_sp = geom.swingarm_pivot_x_mm, geom.swingarm_pivot_height_mm
    x1, y1_0 = X_sp, H_sp
    x2 = X_sp + chain.sprocket_center_x_mm
    y2_0 = H_sp + chain.sprocket_center_y_mm
    IC_x = (y2_0 - y1_0 + m1 * x1 - m2 * x2) / (m1 - m2)
    IC_y = H_sp + m1 * (IC_x - X_sp)
    return IC_x, IC_y


def compute_anti_squat_pct(
    IC_x: float, IC_y: float,
    Y_cg: float, wheelbase: float,
) -> float:
    """
    Anti-Squat % from Foale graphical method.

    Draw a line from rear contact (WB, 0) through IC; find its height
    at the front contact patch vertical (x = 0):

    slope_IC = (0 − IC_y) / (WB − IC_x)
    h_front  = IC_y + slope_IC · (0 − IC_x)
    AS%      = (h_front / Y_cg) · 100        ... Foale Ch. 11 Eq 8.8
    """
    if abs(Y_cg) < 1e-9:
        raise ValueError("Y_cg cannot be zero")
    denom = wheelbase - IC_x
    if abs(denom) < 1e-9:
        raise ValueError("IC_x equals wheelbase — AS line is vertical")
    slope_IC = (0 - IC_y) / denom
    h_front  = IC_y + slope_IC * (0 - IC_x)
    return (h_front / Y_cg) * 100


def compute_anti_squat_swingarm_only(
    swingarm_angle_rad: float,
    wheelbase_mm: float,
    Y_cg_mm: float,
) -> float:
    """
    Anti-squat contribution from swingarm geometry alone (no chain tension).

    When chain tension = 0, the IC retreats to infinity along the swingarm
    axis.  The AS line from rear contact (WB, 0) then runs parallel to the
    swingarm.  Height at x = 0:

        h = −tan(θ_sa) · WB

    For a typical motorcycle θ_sa < 0 → h > 0 → small positive AS%.

    AS_swingarm_only = (−tan(θ_sa) · WB / Y_cg) · 100   ... corrected Eq 8.9

    PREVIOUS (WRONG) formula used tan(θ_sa) · L_sa / Y_cg which:
      a) used swingarm LENGTH instead of WHEELBASE
      b) gave a NEGATIVE result for all typical motorcycles
    This is the correct formula derived from the IC-at-infinity limit.
    """
    if abs(Y_cg_mm) < 1e-9:
        raise ValueError("Y_cg cannot be zero")
    return (-math.tan(swingarm_angle_rad) * wheelbase_mm / Y_cg_mm) * 100


# ── Aggregate static result ───────────────────────────────────────────────────

@dataclass
class StaticResults:
    trail_mm: float
    mechanical_trail_mm: float
    swingarm_angle_deg: float
    X_cg_mm: float
    Y_cg_mm: float
    total_mass_kg: float
    R_front_N: float
    R_rear_N: float
    front_pct: float
    IC_x_mm: float
    IC_y_mm: float
    anti_squat_pct: float
    anti_squat_swingarm_only_pct: float
    chain_contribution_pct: float


def compute_static(
    geom: BikeGeometry,
    components: List[MassComponent],
    chain: ChainGeometry,
) -> StaticResults:
    """Full static analysis — calls all sub-functions in order."""
    trail = compute_trail(geom)
    mt    = compute_mechanical_trail(trail, geom.head_angle_deg)
    theta_sa = compute_swingarm_angle_rad(geom)
    X_cg, Y_cg, total_mass = compute_cog(components)
    R_front, R_rear, front_pct = compute_axle_loads(total_mass, X_cg, geom.wheelbase_mm)
    IC_x, IC_y = compute_instant_centre(geom, chain, theta_sa)
    AS_pct  = compute_anti_squat_pct(IC_x, IC_y, Y_cg, geom.wheelbase_mm)
    AS_sa   = compute_anti_squat_swingarm_only(theta_sa, geom.wheelbase_mm, Y_cg)
    chain_c = AS_pct - AS_sa

    return StaticResults(
        trail_mm=trail,
        mechanical_trail_mm=mt,
        swingarm_angle_deg=math.degrees(theta_sa),
        X_cg_mm=X_cg, Y_cg_mm=Y_cg,
        total_mass_kg=total_mass,
        R_front_N=R_front, R_rear_N=R_rear, front_pct=front_pct,
        IC_x_mm=IC_x, IC_y_mm=IC_y,
        anti_squat_pct=AS_pct,
        anti_squat_swingarm_only_pct=AS_sa,
        chain_contribution_pct=chain_c,
    )
